### ISSUE TITLE
error: expression requires 'long double' type support, but target 'x86_64-unknown-linux-gnu' does not support it

### ISSUE THREAD

**Issue by nathanchance** [ClangBuiltLinux/linux#1497]
After https://github.com/llvm/llvm-project/commit/a8083d42b1c346e21623a1d36d1f0cadd7801d83, I see the following errors with x86_64 allmodconfig

```
../drivers/power/reset/ltc2952-poweroff.c:164:28: error: expression requires  'long double' type support, but target 'x86_64-unknown-linux-gnu' does not support it
        data->wde_interval = 300L * 1E6L;
                                  ^
../drivers/power/reset/ltc2952-poweroff.c:164:21: error: expression requires  'long double' type support, but target 'x86_64-unknown-linux-gnu' does not support it
        data->wde_interval = 300L * 1E6L;
                           ^
../drivers/power/reset/ltc2952-poweroff.c:165:41: error: expression requires  'long double' type support, but target 'x86_64-unknown-linux-gnu' does not support it
        data->trigger_delay = ktime_set(2, 500L*1E6L);
                                               ^
3 errors generated.
```

```
../drivers/usb/dwc2/hcd_queue.c:1744:25: error: expression requires  'long double' type support, but target 'x86_64-unknown-linux-gnu' does not support it
                        delay = ktime_set(0, DWC2_RETRY_WAIT_DELAY);
                                             ^
../drivers/usb/dwc2/hcd_queue.c:62:34: note: expanded from macro 'DWC2_RETRY_WAIT_DELAY'
#define DWC2_RETRY_WAIT_DELAY (1 * 1E6L)
                                 ^
1 error generated.
```

It seems that x87 registers are turned off in `arch/x86/Makefile` with the `-mno-80387` flag so I suppose this error makes sense. Should we just turn these constants into integers (perhaps substituting the `1E6L` with `USEC_PER_SEC`)?

**Comment by nickdesaulniers** [comment]
are these expressions overflowing? In the case of `ktime_set`, the second parameter is an `unsigned long`. `ULONG_MAX` for x86_64 is 4,294,967,295. 1E6L is only 1,000,000.  So looks like no, there is no overflow.

**Comment by nickdesaulniers** [comment]
Is `1E6L` a floating point literal?

**Comment by nathanchance** [comment]
Yes, I had to look that up: https://en.cppreference.com/w/cpp/language/floating_literal

**Comment by nickdesaulniers** [comment]
probably should remove the `L` suffix.

**Comment by nathanchance** [comment]
I don't really see the point of these calculations using floating point values?

`wde_interval` and `trigger_delay` are both `ktime_t`, which is a 64-bit signed integer. `ktime_set()`'s second parameter is an `unsigned long` (integer). I think the `1E6L` can just be turned into `1000000L`, which is `USEC_PER_SEC`.

**Comment by nickdesaulniers** [comment]
> I think the 1E6L can just be turned into 1000000L

Or perhaps just `1E6`.  It's surprising to me to that the `L` suffix means `long double` here and not `long`, which is perhaps the same mistake that the authors made.  I suspect the `E6` suffix changes the meaning of `L` suffixes on literals (which is surprising IMO).

**Comment by nathanchance** [comment]
Yes, the `E6` makes the literal a double: https://godbolt.org/z/cbbKvTWT8

Adding `L` only changes it from `double` to `long double`.

**Comment by nathanchance** [comment]
Patches submitted:

https://lore.kernel.org/r/20211104215047.663411-1-nathan@kernel.org/
https://lore.kernel.org/r/20211104215923.719785-1-nathan@kernel.org/

**Comment by nathanchance** [comment]
v2:

https://lore.kernel.org/r/20211105152049.2522250-1-nathan@kernel.org/
https://lore.kernel.org/r/20211105145802.2520658-1-nathan@kernel.org/

**Comment by nathanchance** [comment]
`drivers/power/reset` patch accepted: https://git.kernel.org/sre/linux-power-supply/c/644106cdb89844be2496b21175b7c0c2e0fab381

**Comment by nathanchance** [comment]
`drivers/usb/dwc2` patch accepted: https://git.kernel.org/gregkh/usb/c/310780e825f3ffd211b479b8f828885a6faedd63

**Comment by nickdesaulniers** [comment]
@phoebewang also fixed this "regression" in LLVM in https://reviews.llvm.org/D114162; since this was a "regression" in ToT LLVM and fixed in ToT LLVM, it doesn't/shouldn't matter whether the kernel patches land or not. Closing.

**Comment by nathanchance** [comment]
Doesn't that only resolve the error for 32-bit x86? These errors all happen with x86_64.

**Comment by phoebewang** [comment]
These errors on x86_64 are intended to match with GCC. https://godbolt.org/z/MhnnPa645
Doesn't GCC fail the same way to the code?

**Comment by nathanchance** [comment]
@phoebewang in the two instances that `clang` caught in the kernel, GCC does not error: https://godbolt.org/z/rfjb5b4rM

GCC does not even warn, which seems like a mistake. I do not really think it is wrong for `clang` to error here (the kernel should not be using a floating point literal) and we already have the patches queued up so I do not really think there is much to do here, unless you think `clang` should try to match GCC more.

**Comment by phoebewang** [comment]
@nathanchance I misunderstood your intention. That's fine. Thanks for your explanation.

**Comment by nathanchance** [comment]
I tested with ToT (33e3554ea33d) and this error is still present. The power supply patch is still pending in Sebastian's tree, Arnd sent a duplicate patch, which will hopefully spur getting this into mainline/-next sooner rather than later: https://lore.kernel.org/r/20211204220249.591482-1-arnd@kernel.org/

**Comment by nathanchance** [comment]
I have pinged the acceptance email asking for a status update, as this is breaking multiple x86_64 configurations and it has been reported on LLVM's issue tracker: https://github.com/llvm/llvm-project/issues/52924

**Comment by nathanchance** [comment]
The second patch has been merged into mainline: https://git.kernel.org/linus/644106cdb89844be2496b21175b7c0c2e0fab381

However, [Linus is unhappy](https://lore.kernel.org/r/CAHk-=wh55h__rC6+RTH7pLLTbnBuuOXPNzGGswVDGN7C6NW1pQ@mail.gmail.com/) with `clang`'s diagnostic in the case of a floating point literal being used only in integer contexts and recommends that `clang` should try and match GCC for compatibility (such as just folding it like GCC does, rather than erroring out: https://godbolt.org/z/rfjb5b4rM).

I am going to leave this open for the moment, until there can be some more discussion around whether or not that is possible (cc @phoebewang). I suspect we won't really run into this again with the kernel but if we do, I suspect we won't have as easy of a time fixing it on the kernel side.

Additionally, that change was not marked for stable but it needs to go there so I will send that along on Monday.

**Comment by phoebewang** [comment]
Thanks @nathanchance for the information.
This is not about error on FP literals when FP is disabled, but error on long double literals when x87 is disabled (and used in macro?). The float/double literals aren't affected https://godbolt.org/z/xYKfTbTTq
Though it's still defective, in practicality, I'd think changing to double literals should solve the problem in most cases.
It should be rare to specifically use long double literals when long double is not support like `qux` in above link.
I'm not expert in front end, cc @asavonic @erichkeane for suggestions on improving it.

**Comment by erichkeane** [comment]
I don't know if there is much we can do.  GCC does constant folding and conversions before some diagnostics, clang does not.  Clang is doing a strict 'correctness' check.

We typically don't want to do what GCC does, as it leads to 'quality of constant folder' issues for diagnostics: that is, a minor regression (or an increase in the depth required for constant folding) will cause errors in a surprising way.

Consider:
500L * 1E6L

as in the example.  Then someone adds a "* SomeFloat" to the end, and it now requires runtime/errors out.  OR the equation is a recursive constant-expression evaluation that manages to hit the implementation limit of frontend folding (say if you were to do 500L * 1E6L * .... *AFewMoreTHings.  Adding a single value would be a 'breaking change' in this way.

**Comment by nickdesaulniers** [comment]
> I saw that and responded to it on the ClangBuiltLinux thread...

Ah, sorry. I have my watch setting for this repo set to "All activity" but I still did not receive a notification of it; looks like I did get an email about it though...huh.

> OR the equation is a recursive constant-expression evaluation that manages to hit the implementation limit of frontend folding (say if you were to do 500L * 1E6L * .... *AFewMoreTHings. Adding a single value would be a 'breaking change' in this way.

Perhaps this diagnostic should use the [`DeferredDiagnosticsEmitter`](https://github.com/llvm/llvm-project/blob/c9f83a8af9dc6d4bd3e864ad5a8df1f441a4b6ef/clang/lib/Sema/Sema.cpp#L1575-L1602) infra, then?

**Comment by erichkeane** [comment]
I don't see how the deferred diagnostics emitter would help.  That is very simply to deal with functions/etc that aren't actually part of the application for the purposes of device compilation (where the 'host' code is simply ignored).  

ETA: If this wasn't "C", I'd say "Use the constexpr engine for this", since that perfectly solves the problem.  But I don't think we have the equivalent of constexpr evaluation available in C mode.

**Comment by nickdesaulniers** [comment]
> I don't see how the deferred diagnostics emitter would help.

For some of the cases described by Linus, clang never emits any floating type to LLVM IR; they're constant folded in clang.  Why error on a type that LLVM never even sees?

**Comment by erichkeane** [comment]
> > I don't see how the deferred diagnostics emitter would help.
> 
> For some of the cases described by Linus, clang never emits any floating type to LLVM IR; they're constant folded in clang. Why error on a type that LLVM never even sees?

In many cases I believe those get folded by the IRBuilder itself during clang-codegen (or, I believe SOMETIMES we do some work in clang-codegen to do other constant folding, but that is after diagnostics).  My understanding of the direction of 'Clang' is that the 'quality' of things like constant-folding shouldn't affect the legality of a program.

Deferred Diagnostics are about deciding entire functions or variables just aren't 'part' of the TU; Note also that this logic STILL happens before the initialization in clang-codegen happens.

  The ONLY thing I could think of (as I mentioned before) would be to find some way to enforce 'constexpr initialization' for some things.  In C++, we have the constexpr keyword to do these.  For C, we could 'try' to always do constant initialization, at the risk of what otherwise seems like flakey initialization.  This is, IMO, bad language design(see my reasoning above), as it depends strongly on the QoI of said constant folder, or expression foldability.

**Comment by nathanchance** [comment]
The `drivers/power/reset/ltc2952-poweroff.c` change is currently in the latest stable release candidates, which should be released in a few days, and the `drivers/usb/dwc2/hcd_queue.c` is already present, so this is fixed as far as stable is concerned as well.

**Comment by nickdesaulniers** [comment]
@gctucker is [reporting](https://github.com/kernelci/kernelci-core/pull/990#issuecomment-1021021122) that some of the 5.4 builds are failing; looks like we need to send backports to stable.

**Comment by nathanchance** [comment]
Both patches are in 5.4 (and every other supported stable release). They need to be merged into `android11-5.4` either via a stable merge or cherry-picked if a merge is not possible because it is a release branch.

https://git.kernel.org/stable/c/9ad421aedc55b4ca062ebc1e62e6518ca8117550
https://git.kernel.org/stable/c/72eb522ae6f1909b84e4323ec7bd6d24d41a8362

**Comment by nathanchance** [comment]
Given this is all squared up on the kernel side, I am closing this. If we want to further discuss making LLVM more compatible with GCC, it is probably worth doing on the LLVM issue tracker.
